### ISSUE TITLE
Separate non-wai specific part out of servant-server

### ISSUE THREAD

**Issue by phadej** [haskell-servant/servant#1166]
This have been discussed, but we have no URL to refer to. Now there's an issue.

**Comment by benkolera** [comment]
Did you have in your plans that you may pull out the Request/Response from client-core and share them with server-core? It looks like that could be a good idea, but I'm unsure at the moment until I really dig in there and check it out.

**Comment by alpmestan** [comment]
Yes, this is an option. We'll have to encode some knowledge about requests and response types somewhere (class methods or the package's own Request/Response types as you suggest) if we want to be able to provide HasServer instances that are backend agnostic and which can easily be "instantiated" to a given backend.

**Comment by benkolera** [comment]
OK, cool. Yeah I didn't know which way that you'd rather go (have a type family and some class methods or have a concrete common datatype that each backend map to their own request/response types). I just noted that servant-client-core went the latter way and latched onto that. 

I'm actually suddenly really booked out now that my Lambda Jam (Australia) workshop has been accepted so this may fall off my radar for the next month. But if you want to write down the design thoughts that were in your head I can run with the leg work when I'm out of conference prep mode. 

I'm super keen to get servant-server-snap to be a first class citizen for my obelisk projects so I'll definitely will get back to it soon. But my brief jump into thinking about the split the other day had me realising that there are a number of design choices to be made here and I should probably ask y'all for what was in your heads already. Saves me putting lots of work into a PR and it not aligning with your vision. 

If telling me your ideas is more difficult than actually just doing the split yourselves let me know and I'll chill out for a bit on it. I can just keep servant-server-snap alive copying and pasting for a little while longer. :slightly_smiling_face:
